src/main.cpp

```diff
@@ -638,19 +638,87 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
     return nEvicted;
 }
 
-bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
+int64_t LockTime(const CTransaction &tx, int flags, const std::vector<int>* prevHeights, const CBlockIndex& block)
 {
-    if (tx.nLockTime == 0)
-        return true;
-    if ((int64_t)tx.nLockTime < ((int64_t)tx.nLockTime < LOCKTIME_THRESHOLD ? (int64_t)nBlockHeight : nBlockTime))
-        return true;
-    BOOST_FOREACH(const CTxIn& txin, tx.vin)
-        if (!txin.IsFinal())
-            return false;
-    return true;
+    assert(prevHeights == NULL || prevHeights->size() == tx.vin.size());
+    int64_t nBlockTime = (flags & LOCKTIME_MEDIAN_TIME_PAST)
+                                    ? block.GetAncestor(std::max(block.nHeight-1, 0))->GetMedianTimePast()
+                                    : block.GetBlockTime();
+
+    bool fEnforceBIP68 = static_cast<uint32_t>(tx.nVersion) >= 2
```

Comment:
Because otherwise you'd be doing a signed comparison and there'd be half
the range of nVersion that wouldn't support BIP 68.

On Tue, Dec 1, 2015 at 4:33 AM, Jorge Timón <notifications@github.com>
wrote:

> In src/main.cpp
> <https://github.com/bitcoin/bitcoin/pull/6312#discussion_r46199226>:
>
> >  {
> > -    if (tx.nLockTime == 0)
> > -        return true;
> > -    if ((int64_t)tx.nLockTime < ((int64_t)tx.nLockTime < LOCKTIME_THRESHOLD ? (int64_t)nBlockHeight : nBlockTime))
> > -        return true;
> > -    BOOST_FOREACH(const CTxIn& txin, tx.vin)
> > -        if (!txin.IsFinal())
> > -            return false;
> > -    return true;
> > +    assert(prevHeights == NULL || prevHeights->size() == tx.vin.size());
> > +    int64_t nBlockTime = (flags & LOCKTIME_MEDIAN_TIME_PAST)
> > +                                    ? block.GetAncestor(std::max(block.nHeight-1, 0))->GetMedianTimePast()
> > +                                    : block.GetBlockTime();
> > +
> > +    bool fEnforceBIP68 = static_cast<uint32_t>(tx.nVersion) >= 2
>
> why is this cast needed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6312/files#r46199226>.
>